src/library/scala/collection/immutable/Map.scala

```diff
@@ -623,7 +623,7 @@ object Map extends ImmutableMapFactory[Map] {
       } else {
         // assert(elems.size == 4)
         elems.get(elem._1) match {
```

Comment:
The change that you mane is better that the original ...but 

It looks to me like the new and old code don't cope with a different value in the map for the same key, and convert to a hashMapBuilder when we could still use a Map4.
Also we could remove the creation of a `Some` via use of a `getOrElse`

something like

in object Map
```
private val flag = new AbstractFunction0[Any, Any] {
  def apply() = flag
}
```
and here 
```
val valueOrFlag = elems.getOrElse(elem._1, Map.flag).asInstanceOf[AnyRef]
if (valueOrFlag eq Map.flag) {
  //its a new key
 convertToHashMapBuilder()
 hashMapBuilder += elem
} else if (valueOrFlag ne elem._2.asInstanceOf[AnyRef]) 
  // its an existing key with a new value 
  elems = elems + elem
//else as key is == and value is eq, its no change
```